### PR TITLE
backend_rights

### PR DESCRIPTION
**Pull request by ulfgebhardt**
## 🍰 Pullrequest

Implements a roles & rights based system for authentication & authorization.

### Issues

- based on #1125 

### Todo

- [x] Admin role check (requires #1121 )
- [ ] Clarify if we want to encode role in jwt to save a constant database query @Mogge 
- [x] test
- [x] set `isAdmin` based on role

### CODE REVIEW

File: backend/src/graphql/resolver/UserResolver.ts [hunk 1]
```diff
@@ -439,6 +475,7 @@ export class UserResolver {
     return new SendPasswordResetEmailResponse(response.data)
   }
 
+  @Authorized([RIGHTS.RESET_PASSWORD])
```
Comment (Mogge) [inline backend/src/graphql/resolver/UserResolver.ts:478]: No token?

File: backend/src/graphql/resolver/UserResolver.ts [hunk 2]
```diff
@@ -565,6 +602,7 @@ export class UserResolver {
     return true
   }
 
+  @Authorized([RIGHTS.CHECK_USERNAME])
```
Comment (Mogge) [inline backend/src/graphql/resolver/UserResolver.ts:605]: No token?

File: backend/src/graphql/resolver/UserResolver.ts [hunk 3]
```diff
@@ -588,6 +626,7 @@ export class UserResolver {
     return true
   }
 
+  @Authorized([RIGHTS.CHECK_EMAIL])
```
Comment (Mogge) [inline backend/src/graphql/resolver/UserResolver.ts:629]: No token

File: backend/src/graphql/resolver/UserResolver.ts [hunk 4]
```diff
@@ -194,6 +195,37 @@ const SecretKeyCryptographyDecrypt = (encryptedMessage: Buffer, encryptionKey: B
...
+      .catch((error) => {
+        throw new Error(error)
+      })
+    user.coinanimation = coinanimation
+    user.isAdmin = true // TODO implement
+    return user
+  }
+  */
+
+  @Authorized([RIGHTS.LOGIN])
```
Comment (Mogge) [inline backend/src/graphql/resolver/UserResolver.ts:228]: No token

**Review by Mogge — CHANGES_REQUESTED** [review]
This is not going to work like this. we have lots of calls that never will have a token, but by setting `@Authorized(SOME_ROLE)` you must have a token, as it is designed by now.

File: backend/src/auth/JWT.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import jwt from 'jsonwebtoken'
+import CONFIG from '../config/'
+import { CustomJwtPayload } from './CustomJwtPayload'
+
+export const decode = (token: string): CustomJwtPayload | null => {
```
[the remark this answers is not in the document]
Comment (Mogge) [reply backend/src/auth/JWT.ts:5]: Stupid question. From the context, of course :)

[the remark this answers is not in the document]
Comment (ulfgebhardt) [reply backend/src/auth/JWT.ts:5]: The JWT is called token in this context I believe.

**Comment by ulfgebhardt** [comment]
@Mogge Please have a look at this - especially the `isAuthorized` - we make a database query on every call to the server to figure out the role - alternatively we can save the role identifier or whatever in JWT and prevent an additional query.

**Review by Mogge — APPROVED** [review]

**Pull request merged**